### PR TITLE
Fix#22

### PR DESCRIPTION
**Pull request by chrstnbwnkl**
- Changed Router base class to be BaseClient base class
- Added Client class derived from BaseClient that uses requests
- Updated all routers to have `client`argument with `Client`as default value

Caveats:

- args passed to router not available via router instance, only through router's client instance (e.g. router.client.proxies instead of router.proxies)

### CODE REVIEW

**Comment by nilsnolde** [comment]
Can you merge master into this branch? Only changed formatting stuff. Also, don't forget to re-install the pre-commit hook with `pre-commit install`.

File: routingpy/base.py [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+        """Must be implemented for inheriting client classes."""
+        pass
+
+    @property
+    def req(self):
+        """Holds the :class:`requests.PreparedRequest` property for the last request."""
+        return self._req
+
+    @staticmethod
+    def _get_body(response):
```
Comment (nilsnolde) [inline routingpy/base.py:164]: put this whole method into the default client, since it has requests specific logic on line 168

File: routingpy/base.py [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+        if 500 <= status_code:
+            raise exceptions.RouterServerError(status_code, body)
+
+        if status_code != 200:
+            raise exceptions.RouterError(status_code, body)
+
+        return body
+
+    @staticmethod
+    def _generate_auth_url(path, params):
```
Comment (nilsnolde) [inline routingpy/base.py:187]: this can stay though, is uses requests but only a static function

File: routingpy/base.py [hunk 3]
```diff
@@ -0,0 +1,206 @@
+# -*- coding: utf-8 -*-
```
Comment (nilsnolde) [inline routingpy/base.py:1]: rename the file to `client_base.py`

File: routingpy/base.py [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+
+        self._req = None
+
+    @abstractmethod
+    def _request(self):
+        """Must be implemented for inheriting client classes."""
+        pass
+
+    @property
+    def req(self):
```
Comment (nilsnolde) [inline routingpy/base.py:159]: I think this can move to the default_client as well

File: routingpy/client_default.py [hunk 5]
```diff
@@ -101,7 +37,7 @@ def __init__(
         retry_timeout=None,
         requests_kwargs=None,
```
Comment (nilsnolde) [inline routingpy/client_default.py:38]: Let's remove this `requests_kwargs` entirely and change the base client to take a `**kwargs` as last argument. makes it more generic. Also remember to change all the "router" instances' `__init__`: in those we can call the argument `**client_kwargs` instead of `**kwargs` to make it more explicit.

I already made peace with releasing a major version update, so we can break existing stuff.

File: routingpy/client_default.py [hunk 6]
```diff
@@ -184,7 +106,7 @@ def _request(
         first_request_time=None,
         retry_counter=0,
         requests_kwargs=None,
```
Comment (nilsnolde) [inline routingpy/client_default.py:108]: I just realized that this is never used in any router method (i.e. no method accepts `requests_kwargs` as arg), so this is actually superfluous. Instead of removing it though, can you put it at the end, call it `**client_kwargs`

File: routingpy/routers/google.py [hunk 7]
```diff
@@ -37,12 +38,13 @@ def __init__(
         requests_kwargs={},
```
Comment (nilsnolde) [inline routingpy/routers/google.py:38]: e.g. here put it at the end of the arg list and call it `**client_kwargs`

**Review by nilsnolde — CHANGES_REQUESTED** [review]
Thanks for the first shot! Left a few todos, call me if anything unclear.

**Comment by nilsnolde** [comment]
smth is not quite right with the linting still.. tons of formatting changes in this PR. don't sweat it though, could also be a mismatch of CI and local stuff: your latest commit passed but it shouldn't have. we'll figure that out after we merged this PR, as long as CI is happy.

File: routingpy/client_default.py (hunk 6 above)
Comment (chrstnbwnkl) [reply routingpy/client_default.py:108]: Actually, it does! See line 193, it's just renamed as `final_requests_kwargs`

Comment (nilsnolde) [reply routingpy/client_default.py:108]: jep but no router actually makes use of it. i.e. if you go into valhalla.directions() there's no `request_kwargs`

Comment (chrstnbwnkl) [reply routingpy/client_default.py:108]: What about the `__init__` of `ORS`? There, the API key is currently passed as part of the `requests_args`.

File: routingpy/routers/graphhopper.py [hunk 8]
```diff
@@ -74,34 +76,37 @@ def __init__(
...
-            >>> print(router.proxies)
+            >>> print(router.client.proxies)
             {'https': '129.125.12.0'}
         :type requests_kwargs: dict
 
         :param retry_over_query_limit: If True, client will not raise an exception
             on HTTP 429, but instead jitter a sleeping timer to pause between
             requests until HTTP 200 or retry_timeout is reached.
-            Default :attr:`routingpy.routers.options.default_retry_over_query_limit`.
+            Default :attr:`routingpy.routers.options.default_over_query_limit`.
```
Comment (nilsnolde) [inline routingpy/routers/graphhopper.py:86]: where do all these unrelated changes come from? also revert

File: routingpy/routers/graphhopper.py [hunk 9]
```diff
@@ -380,7 +385,7 @@ def directions(  # noqa: C901
                     )
 
         return self._parse_directions_json(
-            self._request("/route", get_params=params, dry_run=dry_run), algorithm, elevation
+            self.client._request("/route", get_params=params, dry_run=dry_run), algorithm, elevation
```
Comment (nilsnolde) [inline routingpy/routers/graphhopper.py:388]: you see e.g. here: the `_request` method accepts a client_kwargs that is never passed. that was an oversight by us though: the `directions()` method also doesn't accept a client/requests_kwargs argument. 

The way this generally works is you have 3 hierarchies of "settings" via the kwargs:
1. global
2. per route class instance
3. per single request

While priority increases, i.e. single request > instance > global. A bit overloaded though..

So I propose: let's remove the `client_kwargs` from `_request` entirely and we only let options 1. and 2. active. That's more than enough anyways.

Hope that was clear.

File: routingpy/routers/heremaps.py [hunk 10]
```diff
@@ -74,20 +76,23 @@ def __init__(
...
-            >>> print(router.proxies)
+            >>> print(router.client.proxies)
             {'https': '129.125.12.0'}
         :type requests_kwargs: dict
 
         :param retry_over_query_limit: If True, client will not raise an exception
             on HTTP 429, but instead jitter a sleeping timer to pause between
             requests until HTTP 200 or retry_timeout is reached.
-            Default :attr:`routingpy.routers.options.default_retry_over_query_limit`.
+            Default :attr:`routingpy.routers.options.default_over_query_limit`.
```
Comment (nilsnolde) [inline routingpy/routers/heremaps.py:86]: revert

File: routingpy/routers/mapbox_valhalla.py [hunk 11]
```diff
@@ -62,20 +64,23 @@ def __init__(
...
-            >>> print(router.proxies)
+            >>> print(router.client.proxies)
             {'https': '129.125.12.0'}
         :type requests_kwargs: dict
 
         :param retry_over_query_limit: If True, client will not raise an exception
             on HTTP 429, but instead jitter a sleeping timer to pause between
             requests until HTTP 200 or retry_timeout is reached.
-            Default :attr:`routingpy.routers.options.default_retry_over_query_limit`.
+            Default :attr:`routingpy.routers.options.default_over_query_limit`.
```
Comment (nilsnolde) [inline routingpy/routers/mapbox_valhalla.py:74]: revert

File: routingpy/routers/openrouteservice.py [hunk 12]
```diff
@@ -71,38 +73,41 @@ def __init__(
...
-            >>> print(router.proxies)
+            >>> print(router.client.proxies)
             {'https': '129.125.12.0'}
         :type requests_kwargs: dict
 
         :param retry_over_query_limit: If True, client will not raise an exception
             on HTTP 429, but instead jitter a sleeping timer to pause between
             requests until HTTP 200 or retry_timeout is reached.
-            Default :attr:`routingpy.routers.options.default_retry_over_query_limit`.
+            Default :attr:`routingpy.routers.options.default_over_query_limit`.
```
Comment (nilsnolde) [inline routingpy/routers/openrouteservice.py:83]: revert

File: routingpy/routers/osrm.py [hunk 13]
```diff
@@ -68,30 +70,33 @@ def __init__(
...
-            >>> print(router.proxies)
+            >>> print(router.client.proxies)
             {'https': '129.125.12.0'}
         :type requests_kwargs: dict
 
         :param retry_over_query_limit: If True, client will not raise an exception
             on HTTP 429, but instead jitter a sleeping timer to pause between
             requests until HTTP 200 or retry_timeout is reached.
-            Default :attr:`routingpy.routers.options.default_retry_over_query_limit`.
+            Default :attr:`routingpy.routers.options.default_over_query_limit`.
```
Comment (nilsnolde) [inline routingpy/routers/osrm.py:80]: revert

File: routingpy/routers/valhalla.py [hunk 14]
```diff
@@ -76,32 +78,35 @@ def __init__(
...
-            >>> print(router.proxies)
+            >>> print(router.client.proxies)
             {'https': '129.125.12.0'}
         :type requests_kwargs: dict
 
         :param retry_over_query_limit: If True, client will not raise an exception
             on HTTP 429, but instead jitter a sleeping timer to pause between
             requests until HTTP 200 or retry_timeout is reached.
-            Default :attr:`routingpy.routers.options.default_retry_over_query_limit`.
+            Default :attr:`routingpy.routers.options.default_over_query_limit`.
```
Comment (nilsnolde) [inline routingpy/routers/valhalla.py:88]: revert

**Review by nilsnolde — CHANGES_REQUESTED** [review]
Almost there;) there's quite a few unrelated changes which didn't make sense. best you go through the diff [here on github](https://github.com/gis-ops/routing-py/pull/24/files) and revert all those.

Other than that, three more things:
- please go through the docstrings a final time and update them where you changed things (lots of `request_kwargs` still there)
- let's remove the `client_kwargs` arg to `_request()` entirely. I explained why in the comment.
- the tests are still failing, but all with the same error. See if you can get the reason, else we'll look into tmrw morning together.

**Review by nilsnolde — APPROVED** [review]
found the bug: wasn't your doing. apparently we had a small bug in the tests where it called the wrong URL. and then there was another bug in `responses` which caused our bug to not be detected. What's weird there: I updated the dependencies in the last PR (which must've caused that it failed now), but CI went through.. 

Anyways, I quickly updated that, for me this is good to go. Thanks @chrstnbwnkl
